### PR TITLE
[Core][Spill manager refactor 1] Introduce ObjectSpillManager

### PR DESCRIPTION
**Pull request by scv119**
Create ObjectSpillManager that spills object asynchronously by calling ISpillWorkerPool

Follow ups https://github.com/ray-project/ray/pull/18375 and https://github.com/ray-project/ray/pull/18388

TBD:
add comments and tests.

### CODE REVIEW

File: src/ray/object_manager/plasma/spill_worker_pool.h [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+  virtual void StartSpillWorker(
+      std::vector<ObjectID> objects_to_spill,
+      std::vector<ray::rpc::Address> object_owners,
+      std::function<void(ray::Status, std::vector<std::string>)> spill_callback) = 0;
+
+  virtual void StartDeleteWorker(std::vector<std::string> files_to_delete,
+                                 std::function<void(ray::Status)> delete_callback) = 0;
+};
+}  // namespace plasma
```
Comment (MissiontoMars) [inline src/ray/object_manager/plasma/spill_worker_pool.h:37]: add new line for all files

File: src/ray/object_manager/plasma/object_spill_manager.h [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+  /// \param task_finished_callback The callback once the async spill task finishes.
+  ///   - If the spill task succeed it returns Status::OK() and a map from object_id
+  ///     to the spilled location.
+  ///   - If the spill task failed, it returns failure Status and a map from object_id
+  ///     to empty strings.
+  /// \return wether the spill task is successfully scheduled.
+  ///   - It returns true if the spill task is successfuly scheduled,
+  ///     where the callback is guarantee to be called.
+  ///   - Otherwise it returns false and the callback will not be called
+  bool SubmitSpillTask(std::vector<ObjectID> objects_to_spill,
```
Comment (MissiontoMars) [inline src/ray/object_manager/plasma/object_spill_manager.h:49]:
```suggestion
  bool SubmitSpillTask(cosnt std::vector<ObjectID> &objects_to_spill,
```

File: src/ray/object_manager/plasma/object_spill_manager.h [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+  /// Wether we can submit another spill task.
+  bool CanSubmitSpillTask() const;
+
+  /// Delete a spilled object from external storage.
+  void DeleteSpilledObject(const ObjectID &object_id);
+
+  absl::optional<std::string> GetSpilledUrl(const ObjectID &object_id) const;
+
+ private:
+  struct SpilledObjectURI {
```
Comment (MissiontoMars) [inline src/ray/object_manager/plasma/object_spill_manager.h:63]: Should we using the same name, `URL or `URI.

File: src/ray/object_manager/plasma/object_spill_manager.h [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    std::string file;
+    int64_t offset;
+    int64_t size;
+
+    std::string ToString() const {
+      return absl::StrFormat("%s?offset=%d&size=%d", file, offset, size);
+    }
+  };
+
+  void OnSpillTaskFinished(std::vector<ObjectID> objects_to_spill, ray::Status status,
```
Comment (MissiontoMars) [inline src/ray/object_manager/plasma/object_spill_manager.h:73]:
```suggestion
  void OnSpillTaskFinished(const std::vector<ObjectID> &objects_to_spill, ray::Status status,
```

File: src/ray/object_manager/plasma/object_spill_manager.h [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+                           std::vector<std::string> spilled_urls,
+                           std::function<void(ray::Status status, std::vector<ObjectID>)>
+                               task_finished_callback);
+
+  absl::optional<SpilledObjectURI> ParseSpilledURI(const std::string &s);
+
+  void DeletedSpilledFile(const std::string &spilled_file);
```
Comment (MissiontoMars) [inline src/ray/object_manager/plasma/object_spill_manager.h:80]: Should it be changed to `DeleteSpilledFile` ?

**Comment by scv119** [comment]
@rkooo567 @MissiontoMars this is ready for early feedbacks but missing unit tests and comments.

File: src/ray/object_manager/plasma/spill_worker_pool.h [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+#include "ray/common/id.h"
+#include "src/ray/protobuf/common.pb.h"
+
+namespace plasma {
+class ISpillWorkerPool {
+ public:
+  virtual ~ISpillWorkerPool() = default;
+
+  virtual void StartSpillWorker(
```
Comment (rkooo567) [inline src/ray/object_manager/plasma/spill_worker_pool.h:29]: How do we dispatch requests to workers? (seems like we only have Start*?)

File: src/ray/object_manager/plasma/object_spill_manager.h [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+
+namespace plasma {
+
+class ObjectSpillManager {
+ public:
+  ObjectSpillManager(instrumented_io_context &callback_executor,
+                     std::unique_ptr<ISpillWorkerPool> spill_worker_pool,
+                     int64_t max_active_worker);
+
+  /// Spill a list of objects in an asynchronous task.
```
Comment (rkooo567) [inline src/ray/object_manager/plasma/object_spill_manager.h:37]: `asynchronous task.` sounds a bit confusing. What about calling the method as `SubmitSpillingRequest` and comment as `Submit a request to spill a list of objects`?

**Review by rkooo567 — COMMENTED** [review]
LGTM. Just a few comments. No comment for implementation now as we discussed

**Comment by bveeramani** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've switched our code formatter from YAPF to Black (see #21311).

To prevent issues with merging your code, here's what you'll need to do:

1. Install Black
```
pip install -I black==21.12b0
```

2. Format changed files with Black
```
curl -o format-changed.sh https://gist.githubusercontent.com/bveeramani/42ef0e9e387b755a8a735b084af976f2/raw/7631276790765d555c423b8db2b679fd957b984a/format-changed.sh
chmod +x ./format-changed.sh
./format-changed.sh
rm format-changed.sh
```

3. Commit your changes.
```
git add --all
git commit -m "Format Python code with Black"
```

4. Merge master into your branch.
```
git pull upstream master
```

5. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated `format.sh`.

**Comment by kfstorm** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've updated our formatting configuration for C++ code. (see #22725)

This PR includes C++ code change. To prevent issues with merging your code, here's what you'll need to do:

1. Merge the latest changes from upstream/master branch into your branch.

```bash
git pull upstream master
git merge upstream/master
```

2. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated C++ formatting configuration.

3. Format changed files.

```bash
scripts/format.sh
```

4. Commit your changes.

```bash
git add --all
git commit -m "Format C++ code"
```

**Pull request closed without merging**